zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,248 @@
+//#[allow(missing_docs)]
+/// Errors that can be returned when validating `ValueBalance`s
+enum ValueBalanceError {
+    #[error("value balance contains invalid amounts")]
+    AmountError(#[from] crate::amount::Error),
+}
+
+impl<C> std::ops::Add<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent + rhs.transparent)?,
+            sprout: (value_balance.sprout + rhs.sprout)?,
+            sapling: (value_balance.sapling + rhs.sapling)?,
+            orchard: (value_balance.orchard + rhs.orchard)?,
+        })
+    }
+}
+
+impl<C> std::ops::Sub<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent - rhs.transparent)?,
+            sprout: (value_balance.sprout - rhs.sprout)?,
+            sapling: (value_balance.sapling - rhs.sapling)?,
+            orchard: (value_balance.orchard - rhs.orchard)?,
+        })
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::prelude::*;
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NegativeAllowed> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NonNegative> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn test_operations(
+            value_balance1 in any::<ValueBalance>(),
+            value_balance2 in any::<ValueBalance>())
+        {
```

Comment:
Minor: would it make sense to split this test in two? One for the add operation and one for the sub operation?

Reply:
I thin this is a good idea, i did it in https://github.com/ZcashFoundation/zebra/pull/2505/commits/14f8d23bd75657a28d74e483b7c1422319ba6641